core/src/main/java/org/apache/iceberg/BaseRowDelta.java

```diff
@@ -43,6 +43,10 @@ protected BaseRowDelta self() {
 
   @Override
   protected String operation() {
+    if (!addsDataFiles() && addsDeleteFiles()) {
```

Comment:
If I understand right, wouldn't this condition need to be if (!addsDataFiles() && (addsDeleteFiles() || deletesDataFiles())?

Reply:
Basically I think for a metadata only delete we'd want to make sure the operation is delete

Reply:
yes you're absolutely right. Fixed it. This also needs some more tests, hence why it's WIP

Reply:
@amogh-jahagirdar actually I think `if (!addsDataFiles() && addsDeleteFiles())` is correct as the `RowDelta` API only provides `addRows(DataFile)` and `addDeletes(DeleteFile)`. `deletesDataFiles()` would apply for the `RewriteFiles` API.

Reply:
Ah true, but doesn't the same principle apply for `RewriteFiles` then? I think technically someone can use RewriteFiles to perform just a delete. 

`table.newRewrite().deleteFile(file))`

It's awkward since they could just do a deleteFile but I think it's still possible, and we'd still want to produce the right operation instead of a `REPLACE`. It seems like we should have a check in RewriteFiles as well.

Reply:
Unless you're just saying we'd do RewriteFiles in a separate PR? which I'm good with.

Reply:
yeah I was rather considering doing this in a separate PR

Reply:
@aokolnychyi I  agree that we should use `replace` when it's the same logical set of data but I think the issue is on what `RewriteFiles` actually enforces currently. It doesn't do any validation on this, and I wouldn't expect a logical comparison since that requires a data comparison which isn't feasible.

I wrote the following test in `TestRewriteFiles`:

```
  @TestTemplate
  public void testDelete() {
    commit(table, table.newAppend().appendFile(FILE_A).appendFile(FILE_B), branch);

    table.newRewrite().deleteFile(FILE_A).toBranch(branch).commit();

    table.currentSnapshot();
  }

```
and the operation commits successfully with a REPLACE operation. I know it's strange to use this API (and goes against what's in the JavaDoc) for just deleting a file but as of today it's technically possible and would lead to producing a snapshot operation with the incorrect REPLACE operation.

There's 2 ways to address this incorrectness that I can think of:

1.) Fail at the time of committing if there's only deletes or only additions. If this is the case then we know that the data cannot possibly be logically the same since there'd essentially be a net gain or net loss in data. I'm not sure if there's any other cases but this seems like a reasonable check which doesn't require actually comparing the data.
2.) Don't fail (to preserve the API behavior) and rather at the time of committing, if it's only a delete or only an append, use the correct API to perform the operation with the right summary.

I'm leaning towards 1, since the API docs already makes it clear that from an API perspective `RewriteFiles` should only be used when it's logically the same. If we know for a fact that it's not with a simple check, we can enforce that.

Reply:
I see we have some validation in validateReplacedAndAddedFiles, I think we're just missing this case. I'll raise a PR to show what I mean

Reply:
The use case you highlight is a valid one if FILE_A has all records removed with deletes. Then removing FILE_A has no impact on the table state.

That being said, we should definitely heck if the validation is missing any edge case. Thanks for looking into that!

Reply:
Cool, thanks for explaining! This issue is ultimately orthogonal to the one that is being solved in this PR so we can take that separately.

Reply:
Agreed.